Add update mechanism for Terminfo, and common user-alias data

tecosaur opened the pull request:
Now that we take care of terminfo parsing ourselves, having a clear origin and processing method for arriving at our reference terminfo data seems somewhat important.

The original form was acquired by re-purposing some pre-processed terminfo data from NCurses (I forget the exact source). I considered adding a separate ingestion/processing script, but it occurred to me that it would make sense to have the method for updating the data file be _in_ the data file, by turning it into a quine.

This seems to work rather well, and so long as the NCurses source file format stays the same, updating the terminfo data is now dead simple.

While working on the terminfo files, some minor refactors popped up as "probably nice to have". One of which makes the reported number of flags/numbers/strings actually accurate.

Lastly, to support the ergonomic use of capability variable names instead of the short (read: uninformative) capname, we now also import the NCurses-recognised extended capabilities, and generate/specify some nice aliases for them.

-----

If we separate out the terminfo parser/data into a small stdlib, the state here will be the initial state of the repo.

tecosaur committed:
Turn terminfo_data.jl into a self-updating quine

Is it brilliant or cursed? That's up for debate, but I think it's neat.

This change also revealed some minor discrepancies between the NCurses
source (which we use nwo) and the processed source I originally
used (from some python package? I can't quite recall).

All of the minor discrepancies are cosmetic, with the exception of:
- memory_below moving down by one, and
- backspace_with_bs changing to backspace_if_not_bs

Since we are now using NCurses's source directly, I'm inclined to trust
these changes.

tecosaur committed:
Switch the field ordering of TermCapability

The preferred code to use is the "long" symbol, for readability when
used. I think it makes sense to have it as the first field to reflect
that. While we're at it, we may as well update the field names to better
indicate that "short" is the actual capname, and just call "long"
"name".

tecosaur committed:
Make aliases structurally distinct in TermInfo

tecosaur committed:
Add user-aliases to TermInfo

vtjnash commented on base/terminfo_data.jl:
```diff
@@ -17,524 +32,744 @@ Specification of a single terminal capability.
+"""
+Terminfo extensions that NCurses 6.4 is aware of.
+"""
+const TERM_USER = Dict{Tuple{Type, Symbol}, Tuple{Union{Nothing, Symbol}, String}}(
```
It is typically preferable to keep things in terms of concrete types as much as possible
```suggestion
const TERM_USER = Dict{Tuple{DataType, Symbol}, Union{Tuple{Nothing, String}, Tuple{Symbol, String}}}(
```

vtjnash replied:
Sometimes doing instead a special token value (`:nothing`) rather than using nullable logic might also be simpler? But doesn't seem that necessarily applies here, so no change is actually required (e.g. to make them all `:nothing` instead of `nothing`.)

tecosaur replied:
Made more concrete.

vtjnash reviewed (commented):
SGTM

vtjnash commented on test/terminfo.jl:
```diff
@@ -887,21 +907,21 @@ let
 
 @testset "terminfo" begin
-    dumb = Base.TermInfo(read(IOBuffer(dumb_terminfo), Base.TermInfoRaw))
+    dumb = Terminfo.TermInfo(read(IOBuffer(dumb_terminfo), Terminfo.TermInfoRaw))
```
```
Error in testset terminfo:
2024-02-12 23:03:17 UTC	Error During Test at C:\buildkite-agent\builds\win2k22-amdci6-4\julialang\julia-master\julia-67bbd97d72\share\julia\test\terminfo.jl:909
2024-02-12 23:03:17 UTC	  Got exception outside of a @test
2024-02-12 23:03:17 UTC	  UndefVarError: `Terminfo` not defined in `Main.Test54Main_terminfo`
2024-02-12 23:03:17 UTC	  Stacktrace:
2024-02-12 23:03:17 UTC	    [1] macro expansion
2024-02-12 23:03:17 UTC	      @ C:\buildkite-agent\builds\win2k22-amdci6-4\julialang\julia-master\julia-67bbd97d72\share\julia\test\terminfo.jl:910 [inlined]
```
```suggestion
    dumb = Base.TermInfo(read(IOBuffer(dumb_terminfo), BaseTermInfoRaw))
```

tecosaur replied:
Ooops, this slipped in by accident while experimenting with "make terminfo handling a package".

vtjnash commented on test/terminfo.jl:
```diff
@@ -887,21 +907,21 @@ let
     end
 
-    xterm = Base.TermInfo(read(IOBuffer(xterm_terminfo), Base.TermInfoRaw))
+    xterm = Terminfo.TermInfo(read(IOBuffer(xterm_terminfo), Terminfo.TermInfoRaw))
```
```suggestion
    xterm = Base.TermInfo(read(IOBuffer(xterm_terminfo), Base.TermInfoRaw))
```

tecosaur committed:
Change terminfo update logic, and update

Using the tagged releases in the ncurses repository isn't actually a
good idea, since it fails to capture the patch releases made
subsequently.

Instead, we switch to grabbing the latest version and recording which
patch release it is. Normally, there would be a danger of grabbing
un-released changes, however this doesn't seem like an actual risk in
this case given how the ncurses repo seems to operate.

tecosaur commented:
I noticed that the NCurses release tags weren't quite right, so have tweaked the update logic, and updated again (so now we have the latest 6.4 release, not the initial 6.4 release).

Also, there are enough commits here worth keeping IMO that I'm going to slap a "don't squash" label on this.
